### PR TITLE
replace refs to values already in the session

### PR DESCRIPTION
**Pull request by snopoke**
http://manage.dimagi.com/default.asp?236750

@calellowitz 
cc @phillipm

### CODE REVIEW

File: corehq/apps/app_manager/tests/test_form_workflow.py [hunk 1]
```diff
@@ -136,6 +140,46 @@ def test_manual_form_link(self):
...
         self.assertXmlPartialEqual(self.get_xml('form_link_tdh'), factory.app.create_suite(), "./entry")
 
+    def test_reference_to_missing_session_variable_in_stack(self):
+        # http://manage.dimagi.com/default.asp?236750
+        #
+        # Stack create blocks do not update the session after each datum
+        # so items put into the session in one step aren't available later steps
+        #
+        #    <datum id="case_id_A" value="instance('commcaresession')/session/data/case_id_new_A"/>
+        # -  <datum id="case_id_B" value="instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_A]/index/host"/>
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_form_workflow.py]: E501 line too long (147 > 115 characters)

File: corehq/apps/app_manager/tests/test_form_workflow.py [hunk 2]
```diff
@@ -136,6 +140,46 @@ def test_manual_form_link(self):
...
+        # +  <datum id="case_id_B" value="instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_new_A]/index/host"/>
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_form_workflow.py]: E501 line too long (151 > 115 characters)

File: corehq/apps/app_manager/tests/test_form_workflow.py [hunk 3]
```diff
@@ -136,6 +140,46 @@ def test_manual_form_link(self):
...
+        factory.form_opens_case(m1f0, case_type='episode', is_subcase=True, is_extension=True)
+
+        m2, m2f0 = factory.new_advanced_module('tests', 'episode')
+        factory.form_requires_case(m2f0, 'episode')
+        factory.advanced_form_autoloads(m2f0, AUTO_SELECT_CASE, 'host', 'load_episode_0')
+
+        m1f0.post_form_workflow = WORKFLOW_FORM
+        m1f0.form_links = [
+            FormLink(xpath="true()", form_id=m2f0.unique_id, datums=[
+                FormDatum(name='case_id_load_episode_0', xpath="instance('commcaresession')/session/data/case_id_new_episode_0")
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_form_workflow.py]: E501 line too long (128 > 115 characters)

File: corehq/apps/app_manager/tests/test_form_workflow.py [hunk 4]
```diff
@@ -298,3 +342,32 @@ def test_form_workflow_module_in_root(self):
...
         self.assertXmlPartialEqual(self.get_xml('suite-workflow-root'), app.create_suite(), "./entry")
+
+
+class TestReplaceSessionRefs(SimpleTestCase):
+    def test_replace_session_references_in_stack(self):
+        children = [
+            CommandId('m0'),
+            StackDatum(id='a', value=session_var('new_a')),
+            StackDatum(id='b', value=session_var('new_b')),
+            StackDatum(id='c', value="instance('casedb')/case/[@case_id = {a}]/index/parent".format(a=session_var('a'))),
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_form_workflow.py]: E501 line too long (121 > 115 characters)

**Comment by czue** [comment]
build is failing
```
======================================================================
FAIL: * Register house (case type = house, basic)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_case_list_form.py", line 161, in test_case_list_form_parent_child_basic
    self.assertXmlEqual(self.get_xml('case-list-form-suite-parent-child-basic'), factory.app.create_suite())
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 32, in assertXmlEqual
    _check_shared(expected, actual, LXMLOutputChecker(), "xml")
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 126, in _check_shared
    raise AssertionError(message)
AssertionError: XML mismatch
--- want.xml
+++ got.xml
@@ -183,7 +183,7 @@
     <stack>
       <create if="count(instance('commcaresession')/session/data/return_to) = 1 and instance('commcaresession')/session/data/return_to = 'm2' and count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_new_person_0]) &gt; 0">
         <command value="'m2'"/>
-        <datum id="parent_id" value="instance('commcaresession')/session/data/case_id_new_person_0"/>
+        <datum id="parent_id" value="instance('commcaresession')/session/data/case_id"/>
         <datum id="case_id" value="instance('commcaresession')/session/data/case_id_new_person_0"/>
       </create>
       <create if="count(instance('commcaresession')/session/data/return_to) = 1 and instance('commcaresession')/session/data/return_to = 'm2' and count(instance('casedb')/casedb/case[@case_id=instance('commcaresession')/session/data/case_id_new_person_0]) = 0">
----------------------------------------------------------------------
```
```
======================================================================
FAIL: test_link_to_child_module (corehq.apps.app_manager.tests.test_form_workflow.TestFormWorkflow)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/test_form_workflow.py", line 111, in test_link_to_child_module
    self.assertXmlPartialEqual(self.get_xml('form_link_tdh'), factory.app.create_suite(), "./entry")
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 26, in assertXmlPartialEqual
    self.assertXmlEqual(expected, actual, normalize=False)
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 32, in assertXmlEqual
    _check_shared(expected, actual, LXMLOutputChecker(), "xml")
  File "/mnt/commcare-hq-ro/corehq/apps/app_manager/tests/util.py", line 126, in _check_shared
    raise AssertionError(message)
AssertionError: XML mismatch
--- want.xml
+++ got.xml
@@ -36,7 +36,7 @@
         <datum id="case_id" value="instance('commcaresession')/session/data/case_id"/>
         <datum id="case_id_new_visit_0" value="uuid()"/>
         <command value="'m2'"/>
-        <datum id="case_id_load_visit_0" value="uuid()"/>
+        <datum id="case_id_load_visit_0" value="instance('commcaresession')/session/data/case_id_new_visit_0"/>
         <command value="'m2-f0'"/>
       </create>
     </stack>
----------------------------------------------------------------------
```

File: corehq/apps/app_manager/tests/test_case_list_form.py [hunk 5]
```diff
@@ -161,16 +157,14 @@ def test_case_list_form_parent_child_basic(self):
...
-          * Update person (case type = person, case list form = 'Register person form', basic, parent module = 'Update house')
-              * Update person form
-        """
+        # * Register house (case type = house, basic)
+        #   * Register house form
+        # * Register person (case type = person, parent select = 'Register house', basic)
+        #   * Register person form
+        # * Update house (case type = house, case list form = 'Register house')
+        #   * Update house form
+        #   * Update person (case type = person, case list form = 'Register person form', basic, parent module = 'Update house')
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_case_list_form.py]: E501 line too long (128 > 115 characters)

File: corehq/apps/app_manager/tests/test_case_list_form.py [hunk 6]
```diff
@@ -198,16 +192,14 @@ def test_case_list_form_parent_child_submodule_basic(self):
...
-          * Update person (case type = person, case list form = 'Register person form', advanced, parent module = 'Update house')
-              * Update person form
-        """
+        # * Register house (case type = house, basic)
+        #   * Register house form
+        # * Register person (case type = person, parent select = 'Register house', advanced)
+        #   * Register person form
+        # * Update house (case type = house, case list form = 'Register house')
+        #   * Update house form
+        #   * Update person (case type = person, case list form = 'Register person form', advanced, parent module = 'Update house')
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_case_list_form.py]: E501 line too long (131 > 115 characters)

File: corehq/apps/app_manager/tests/test_case_list_form.py [hunk 7]
```diff
@@ -276,16 +267,14 @@ def test_case_list_form_parent_child_submodule_advanced_rename_case_var(self):
...
```
Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_case_list_form.py]: E501 line too long (131 > 115 characters)

Comment (dimagimon) [inline corehq/apps/app_manager/tests/test_case_list_form.py]: E501 line too long (131 > 115 characters)

**Comment by snopoke** [comment]
:white_check_mark:

**Comment by snopoke** [comment]
:white_check_mark: bump

File: corehq/apps/app_manager/suite_xml/post_process/workflow.py [hunk 8]
```diff
@@ -529,4 +547,40 @@ def __ne__(self, other):
         return not self == other
 
     def __repr__(self):
-        return 'DatumMeta(id={}, case_type={})'.format(self.id, self.case_type)
+        return 'WorkflowDatumMeta(id={}, case_type={})'.format(self.id, self.case_type)
```
Comment (benrudolph) [inline corehq/apps/app_manager/suite_xml/post_process/workflow.py]: nit i think this'll crash on a unicode case_type if we allow that

File: corehq/apps/app_manager/tests/test_case_list_form.py [hunk 9]
```diff
@@ -276,16 +267,14 @@ def test_case_list_form_parent_child_submodule_advanced_rename_case_var(self):
...
+        #       * Update person form
```
Comment (benrudolph) [inline corehq/apps/app_manager/tests/test_case_list_form.py]: just curious, any reason for the switch to `# `

Comment (benrudolph) [inline corehq/apps/app_manager/tests/test_case_list_form.py]: nvm, i see tests were failing

**Pull request merged**